File: src/main/java/com/ggardet/core/events/EventPublications.java
package com.ggardet.core.events;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.modulith.events.CompletedEventPublications;
import org.springframework.modulith.events.IncompleteEventPublications;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;

import static java.time.Duration.ofSeconds;

@Component
@RequiredArgsConstructor
@Slf4j
class EventPublications {
    private final IncompleteEventPublications incompleteEventPublications;
    private final CompletedEventPublications completedEventPublications;

    @Scheduled(fixedRate = 60_000, initialDelay = 1_000)
    void reSubmitIncompleteEventsOlderThan60s() {
        log.info("Resubmit incomplete publications");
        incompleteEventPublications.resubmitIncompletePublicationsOlderThan(ofSeconds(60));
    }

    @Scheduled(fixedRate = 60_000, initialDelay = 1_000)
    void listCompleteEvents() {
        final int numberOfPublications = completedEventPublications.findAll().size();
        log.info("List of {} complete publications", numberOfPublications);
        completedEventPublications.findAll().forEach(event -> {
            log.info("Event Id={}", event.getIdentifier());
            log.info("Event={}", event.getEvent());
            log.info("Publication Date={}", event.getPublicationDate());
            log.info("Completion Date={}", event.getCompletionDate());
        });
    }
}

File: src/main/kotlin/com/ggardet/core/events/EventPublications.kt
package com.ggardet.core.events

import org.slf4j.LoggerFactory
import org.springframework.context.ApplicationEventPublisher
import org.springframework.modulith.events.CompletedEventPublications
import org.springframework.modulith.events.IncompleteEventPublications
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Component
import java.time.Duration.ofSeconds

@Component
class EventPublications(
    private val eventPublisher: ApplicationEventPublisher,
    private val incompleteEventPublications: IncompleteEventPublications,
    private val completedEventPublications: CompletedEventPublications
) {
    
    private val log = LoggerFactory.getLogger(EventPublications::class.java)

    @Scheduled(fixedRate = 60_000, initialDelay = 1_000)
    fun reSubmitIncompleteEventsOlderThan60s() {
        log.info("Resubmit incomplete publications")
        incompleteEventPublications.resubmitIncompletePublicationsOlderThan(ofSeconds(60))
    }

    @Scheduled(fixedRate = 60_000, initialDelay = 1_000)
    fun listCompleteEvents() {
        val numberOfPublications = completedEventPublications.findAll().size
        log.info("List of {} complete publications", numberOfPublications)
        completedEventPublications.findAll().forEach { event ->
            log.info("Event Id={}", event.identifier)
            log.info("Event={}", event.event)
            log.info("Publication Date={}", event.publicationDate)
            log.info("Completion Date={}", event.completionDate)
        }
    }
    
    fun publishEvent(event: Any) {
        eventPublisher.publishEvent(event)
    }
}